app/models/targeting.rb

```diff
@@ -36,9 +36,13 @@ def resolve_hosts!
     raise ::Foreman::Exception, _('Cannot resolve hosts without a bookmark or search query') if bookmark.nil? && search_query.blank?
 
     self.search_query = bookmark.query if dynamic? && bookmark.present?
-    self.save!
-    self.touch(:resolved_at)
-    self.hosts = User.as(user.login) { Host.authorized(RESOLVE_PERMISSION, Host).search_for(search_query) }
+    self.resolved_at = Time.zone.now
+    self.validate!
+    # avoid validation of hosts objects - tey will be loaded for no reason.
+    host_ids = User.as(user.login) { Host.authorized(RESOLVE_PERMISSION, Host).search_for(search_query).pluck(:id) }
+    # this can be optimized even more, by introducing bulk insert
+    self.targeting_hosts.build(host_ids.map { |id| { :host_id => id } })
```

Comment:
No, since `TargetingHost` is a model by itself, it would have a different id.
You mean `self.host_ids = host_ids` probably (assign the array to the property that should hold ids only). I have tried it, ActiveRecord tries to validate that those hosts exist and puts them into `self.hosts` property under the hood, so it queries the database with `SELECT hosts.* from hosts where id in (...)` which I want to avoid, since there is no use for hosts here.